### ISSUE TITLE
Make CRD namespaced

### ISSUE THREAD

**Issue by cliveseldon** [SeldonIO/seldon-core#95]
Should the CRD definition be namespaced so it and the operator that uses it are all in one namespace instead of cluster-wide?

**Comment by Raab70** [comment]
What's weird about this is that the CRD is scope `Namespaced` in the spec. I was able to get another `seldon-core` standing in a neighboring namespace by adding a flag to turn off the CRD definition and by fixing the hardcoded default namespaces for RBAC. After doing this I did confirm that the CRD is unavailable when trying to deploy a model to another namespace, but we're still getting this collision. 

I also tried from there making the scope `Cluster` and deploying a model to another namespace. Which worked! But the pod was running in the `default` namespace... 

This leads me to believe maybe theres some `default` namespaces hardcoded elsewhere that I missed?

**Comment by cliveseldon** [comment]
That sounds like great progress. If the Deployment is created in the default namespace there must be something missing in the clusterManager java code for namespacing or how its being set.

https://github.com/SeldonIO/seldon-core/blob/7a1d6144cc1f70554f64fbe35569829319e97e9d/cluster-manager/src/main/java/io/seldon/clustermanager/k8s/SeldonDeploymentControllerImpl.java#L172

Are you sure your SeldonDeployment has a namespace value?

**Comment by Raab70** [comment]
It appears it does not...
```$ kubectl get sdep demo-model2-example -o yaml
apiVersion: machinelearning.seldon.io/v1alpha1
kind: SeldonDeployment
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"machinelearning.seldon.io/v1alpha1","kind":"SeldonDeployment","metadata":{"annotations":{},"labels":{"app":"seldon"},"name":"demo-model2-example","namespace":"demomodel"},"spec":{"annotations":{"deployment_version":"0.1","project_name":"Demo Model2"},"name":"demo-model2-deployment","oauth_key":"oauth-key","oauth_secret":"oauth-secret","predictors":[{"annotations":{"predictor_version":"0.1"},"componentSpec":{"spec":{"containers":[{"image":"gcr.io/i-amlg-dev/cbsiamlg/demomodel:master","imagePullPolicy":"IfNotPresent","name":"demo-model2","resources":{"requests":{"memory":"1Mi"}}}],"terminationGracePeriodSeconds":20}},"graph":{"children":[],"endpoint":{"type":"REST"},"name":"demo-model2","type":"MODEL"},"name":"demo-model2-predictor","replicas":1}]}}
  clusterName: ""
  creationTimestamp: 2018-02-14T23:23:14Z
  deletionGracePeriodSeconds: null
  deletionTimestamp: null
  initializers: null
  labels:
    app: seldon
  name: demo-model2-example
  namespace: ""
  resourceVersion: "1413523"
  selfLink: /apis/machinelearning.seldon.io/v1alpha1/demo-model2-example
  uid: 07a980bf-11de-11e8-96c5-42010a8000ca
spec:
  annotations:
    deployment_version: "0.1"
    project_name: Demo Model2
  name: demo-model2-deployment
  oauth_key: oauth-key
  oauth_secret: oauth-secret
  predictors:
  - annotations:
      predictor_version: "0.1"
    componentSpec:
      spec:
        containers:
        - image: gcr.io/i-amlg-dev/cbsiamlg/demomodel:master
          imagePullPolicy: IfNotPresent
          name: demo-model2
          resources:
            requests:
              memory: 1Mi
        terminationGracePeriodSeconds: 20
    graph:
      children: []
      endpoint:
        type: REST
      name: demo-model2
      type: MODEL
    name: demo-model2-predictor
    replicas: 1```

Very strange, I can't seem to get it to set. I have tried:
1. `kubectl apply --namespace demomodel -f deployment.json`  
1. setting up a context and it doesn't get passed in. 
1. Specifying it in the `deployment.json`

This deployment: 
```{
    "apiVersion": "machinelearning.seldon.io/v1alpha1",
    "kind": "SeldonDeployment",
    "metadata": {
        "labels": {
            "app": "seldon"
        },
        "name": "demo-model2-example",
        "namespace": "demomodel"
    },
"spec": {
    "annotations": {
        "project_name": "Demo Model2",
        "deployment_version": "0.1"
        },
        "name": "demo-model2-deployment",
        "oauth_key": "oauth-key",
        "oauth_secret": "oauth-secret",
        "predictors": [
            {
                "componentSpec": {
                    "spec": {
                        "containers": [
                            {
                                "image": "gcr.io/i-amlg-dev/cbsiamlg/demomodel:master",
                                "imagePullPolicy": "IfNotPresent",
                                "name": "demo-model2",
                                "resources": {
                                    "requests": {
                                        "memory": "1Mi"
                                    }
                                }
                            }
                        ],
                        "terminationGracePeriodSeconds": 20
                    }
                },
                "graph": {
                    "children": [],
                    "name": "demo-model2",
                    "endpoint": {
                        "type": "REST"
                    },
                    "type": "MODEL"
                },
                "name": "demo-model2-predictor",
                "replicas": 1,
                "annotations": {
                    "predictor_version": "0.1"
                }
            }
        ]
    }
}```

produces this:
```
$ kubectl get sdep demo-model2-example -o yaml
apiVersion: machinelearning.seldon.io/v1alpha1
kind: SeldonDeployment
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"machinelearning.seldon.io/v1alpha1","kind":"SeldonDeployment","metadata":{"name":"demo-model2-example","namespace":"","selfLink":"/apis/machinelearning.seldon.io/v1alpha1/demo-model2-example","uid":"07a980bf-11de-11e8-96c5-42010a8000ca","generation":0,"creationTimestamp":"2018-02-14T23:23:14Z","labels":{"app":"seldon"},"clusterName":""},"spec":{"name":"demo-model2-deployment","predictors":[{"name":"demo-model2-predictor","graph":{"name":"demo-model2","type":"MODEL","endpoint":{"service_host":"0.0.0.0","service_port":9000,"type":"REST"}},"componentSpec":{"metadata":{"labels":{"seldon-app":"demo-model2-deployment"}},"spec":{"containers":[{"name":"demo-model2","image":"gcr.io/i-amlg-dev/cbsiamlg/demomodel:master","ports":[{"name":"http","containerPort":9000}],"env":[{"name":"PREDICTIVE_UNIT_SERVICE_PORT","value":"9000"},{"name":"PREDICTIVE_UNIT_PARAMETERS","value":"[]"},{"name":"PREDICTIVE_UNIT_ID","value":"demo-model2"},{"name":"PREDICTOR_ID","value":"demo-model2-predictor"},{"name":"SELDON_DEPLOYMENT_ID","value":"demo-model2-example"}],"resources":{"requests":{"memory":"1Mi"}},"livenessProbe":{"handler":{"tcpSocket":{"port":"http"}},"initialDelaySeconds":10,"periodSeconds":5},"readinessProbe":{"handler":{"tcpSocket":{"port":"http"}},"initialDelaySeconds":10,"periodSeconds":5},"lifecycle":{"preStop":{"exec":{"command":["/bin/sh","-c","/bin/sleep 5"]}}},"imagePullPolicy":"IfNotPresent"}],"terminationGracePeriodSeconds":20}},"replicas":1,"annotations":{"predictor_version":"0.1"}}],"oauth_key":"oauth-key","oauth_secret":"oauth-secret","annotations":{"deployment_version":"0.1","project_name":"Demo Model2"}},"status":{"predictorStatus":[{"name":"demo-model2-deployment-demo-model2-predictor","replicas":1,"replicasAvailable":1}]}}
  clusterName: ""
  creationTimestamp: 2018-02-14T23:23:14Z
  generation: 0
  initializers: null
  labels:
    app: seldon
  name: demo-model2-example
  namespace: ""
  resourceVersion: "1413950"
  selfLink: /apis/machinelearning.seldon.io/v1alpha1/demo-model2-example
  uid: 07a980bf-11de-11e8-96c5-42010a8000ca
spec:
  annotations:
    deployment_version: "0.1"
    project_name: Demo Model2
  name: demo-model2-deployment
  oauth_key: oauth-key
  oauth_secret: oauth-secret
  predictors:
  - annotations:
      predictor_version: "0.1"
    componentSpec:
      metadata:
        labels:
          seldon-app: demo-model2-deployment
      spec:
        containers:
        - env:
          - name: PREDICTIVE_UNIT_SERVICE_PORT
            value: "9000"
          - name: PREDICTIVE_UNIT_PARAMETERS
            value: '[]'
          - name: PREDICTIVE_UNIT_ID
            value: demo-model2
          - name: PREDICTOR_ID
            value: demo-model2-predictor
          - name: SELDON_DEPLOYMENT_ID
            value: demo-model2-example
          image: gcr.io/i-amlg-dev/cbsiamlg/demomodel:master
          imagePullPolicy: IfNotPresent
          lifecycle:
            preStop:
              exec:
                command:
                - /bin/sh
                - -c
                - /bin/sleep 5
          livenessProbe:
            handler:
              tcpSocket:
                port: http
            initialDelaySeconds: 10
            periodSeconds: 5
          name: demo-model2
          ports:
          - containerPort: 9000
            name: http
          readinessProbe:
            handler:
              tcpSocket:
                port: http
            initialDelaySeconds: 10
            periodSeconds: 5
          resources:
            requests:
              memory: 1Mi
        terminationGracePeriodSeconds: 20
    graph:
      endpoint:
        service_host: 0.0.0.0
        service_port: 9000
        type: REST
      name: demo-model2
      type: MODEL
    name: demo-model2-predictor
    replicas: 1
status:
  predictorStatus:
  - name: demo-model2-deployment-demo-model2-predictor
    replicas: 1
    replicasAvailable: 1
```

**Comment by cliveseldon** [comment]
I think there is an issue in the operator which is presently hardcoded to watch for new seldondeployments in the "default" namespace.

So I think there are changes required to
  *  make the CRD namespaced
  *  make the operator look for CRs created in its namespace

**Comment by Raab70** [comment]
The CRD is namespaced but there is definitely a bug in deploying models to a namespace besides default. I was able to get `seldon-core` installed into the default namespace and another namespace as I mentioned above and I created two deployments:
```$ kubectl get sdep demo-model-example -o yaml | grep namespace:
  namespace: default
$ kubectl get sdep --namespace demomodel demo-model2-example -o yaml | grep namespace:
  namespace: demomodel
```
But the model never starts:
```
NAMESPACE     NAME                                                         READY     STATUS    RESTARTS   AGE
default       demo-model-deployment-demo-model-predictor-9c5dff6fd-grlb9   2/2       Running   0          2m
default       redis-df886d999-hdgwh                                        1/1       Running   0          7m
default       seldon-cluster-manager-75db85f5f9-bdxgg                      1/1       Running   0          7m
demomodel     redis-df886d999-dvckl                                        1/1       Running   0          2m
demomodel     seldon-cluster-manager-576c6ddf5-h9mj6                       1/1       Running   0          2m```

I'll take a look at the operator and try to find where it's looking for CRs. Any guidance is appreciated.

**Comment by cliveseldon** [comment]
Yep. We are working on a fix to make the CRD and the Operator fully namespaced.

**Comment by cliveseldon** [comment]
Fixed by #97 
  * The CRD is in a separate helm script which needs to be installed first and then you can install seldon-core into different namespaces. 
 * The hardwired nodeports have been removed

See the example notebooks
